cdap-watchdog/src/main/java/io/cdap/cdap/logging/appender/CompositeLogAppender.java

```diff
@@ -0,0 +1,120 @@
+
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(CompositeLogAppender.class);
+  private final List<LogAppender> appenders;
+
+  /**
+   * Constructs a CompositeLogAppender with the given list of appenders.
+   *
+   * @param appenders the list of {@link LogAppender} instances to delegate logging events to
+   */
+  public CompositeLogAppender(List<LogAppender> appenders) {
+    if (appenders == null) {
+      throw new IllegalArgumentException("Appenders list cannot be null");
+    }
+
+    // Make the appenders list unmodifiable to ensure thread safety and prevent external
+    // modifications.
+    this.appenders = Collections.unmodifiableList(new ArrayList<>(appenders));
+    setName(getClass().getName());
+  }
+
+  @Override
+  public void start() {
+    appenders.forEach(appender -> safelyExecute(
+        appender::start,
+        String.format("Failed to start appender: %s", appender.getName())
+    ));
+    super.start();
+  }
+
+  @Override
+  public void stop() {
+    super.stop();
+    appenders.forEach(appender -> safelyExecute(
+        appender::stop,
+        String.format("Failed to stop appender: %s", appender.getName())
+    ));
+  }
+
+  @Override
+  protected void appendEvent(LogMessage logMessage) {
+    appenders.forEach(appender -> safelyExecute(
+        () -> appender.appendEvent(logMessage),
+        String.format("Failed to append log message to appender: %s. Message: %s",
+            appender.getName(), logMessage.getFormattedMessage())
+    ));
+  }
+
+  @Override
+  public void setContext(Context context) {
+    super.setContext(context);
+    appenders.forEach(appender -> safelyExecute(
+        () -> appender.setContext(context),
+        String.format("Failed to set context to appender: %s", appender.getName())
+    ));
+  }
+
+  @Override
+  public synchronized void doAppend(ILoggingEvent eventObject) {
+    if (shouldSkipLogging()) {
+      return;
+    }
+    super.doAppend(eventObject);
+  }
+
+  @VisibleForTesting
+  boolean shouldSkipLogging() {
+    // If any of the LogAppender is LocalLogAppender then, ignore logs coming from the log process
+    // pipeline, otherwise it'll become an infinite loop of logs.
+    // This won't guard against the case that an appender starts a new thread and emit log per
+    // event (something like what this class does). If that's the case, the appender itself need to
+    // guard against it, similar to what's being done in here.
+    // They are still logged via other log appender (e.g. log to cdap.log), but just not being
+    // collected via the log collection system.
+    return appenders.stream()
+        .filter(LocalLogAppender.class::isInstance)
+        .map(LocalLogAppender.class::cast)
+        .anyMatch(localAppender -> localAppender.getPipelineThreads()
+            .get()
+            .contains(Thread.currentThread()));
+  }
+
+  private void safelyExecute(Runnable action, String errorMessage) {
```

Comment:
The term `safety` can be confusing. Since you want to perform action and not throw exception. You can rename it something like `executeNoThrow()` or something better.

How did the caller handle the exceptions previously? Did it use to retry or ignore them? If it was retried, then the new behavior is different.